src/store.ts

```diff
@@ -0,0 +1,823 @@
+  timeout: number;
+  param: any;
+  value: any;
+  tags: Record<string, string>;
+  iKeyForCreate?: string;
+  iKeyForComplete?: string;
+  createdOn?: number;
+  completedOn?: number;
+  callbacks?: Map<string, CallbackData>;
+}
+
+interface CallbackData {
+  id: string;
+  type: "resume" | "notify";
+  promiseId: string;
+  rootPromiseId: string;
+  recv: string;
+  timeout: number;
+  createdOn: number;
+}
+
+interface TaskData {
+  id: string;
+  counter: number;
+  state: "init" | "enqueued" | "claimed" | "completed";
+  type: "invoke" | "resume" | "notify";
+  recv: string;
+  rootPromiseId: string;
+  leafPromiseId: string;
+  pid?: string;
+  ttl?: number;
+  expiry?: number;
+  createdOn: number;
+  completedOn?: number;
+}
+
+interface Router {
+  route(promise: DurablePromiseData): any;
+}
+class TagRouter implements Router {
+  private tag: string;
+
+  constructor(tag: string = "resonate:invoke") {
+    this.tag = tag;
+  }
+
+  route(promise: DurablePromiseData): any {
+    return promise.tags?.[this.tag];
+  }
+}
+
+export class Server {
+  private promises: PromiseStore;
+  private tasks: TaskStore;
+  private timeouts: Heapq<number>;
+  private routers: Array<Router>;
+  private targets: Record<string, string>;
+
+  constructor() {
+    this.promises = new PromiseStore();
+    this.tasks = new TaskStore();
+    this.timeouts = new Heapq();
+    this.routers = new Array(new TagRouter());
+    this.targets = { default: "local://any@defaul" };
+  }
+
+  next(): [() => void, number] {
+    return [this.step, this.timeouts.top()];
+  }
+
+  step() {
+    const time = Date.now();
+
+    for (const promise of this.promises.scan()) {
+      if (promise.state === "pending" && time >= promise.timeout) {
+        var applied = this.transition(promise.id, "rejected_timedout")[2];
+        util.assert(applied, "expected promise to be timedout");
+      }
+    }
+
+    // transition expired tasks to init
+    for (const task of this.tasks.scan()) {
+      if (["enqueued", "claimed"].includes(task.state)) {
+        if (task.expiry === undefined) {
+          throw new Error("task must have an expiry");
+        }
+        util.assert(task.expiry !== undefined);
+
+        if (time >= task.expiry) {
+          var [_, applied] = this.tasks.transition(
+            task.id,
+            "init",
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            true,
+          );
+          util.assert(applied);
+        }
+      }
+
+      //
+    }
+  }
+
+  private transition(
+    id: string,
+    to: "pending" | "resolved" | "rejected" | "rejected_canceled" | "rejected_timedout",
+    strict?: boolean,
+    timeout?: number,
+    ikey?: string,
+    value?: any,
+    tags?: Record<string, string>,
+  ): [DurablePromiseData, TaskData | undefined, boolean] {
```

Comment:
```suggestion
  ): {promise:DurablePromiseData, task:TaskData | undefined, applied:boolean} {
```
Same as above, use an object instead of a tuple.